arangod/Aql/MaterializeExecutor.cpp

```
@@ -37,58 +37,62 @@
+    : MaterializeExecutorBase(infos),
+      _collection(infos.collection()->getCollection()->getPhysical()) {
+  TRI_ASSERT(_collection != nullptr);
 }
 
-template<bool localDocumentId>
-MaterializeExecutor<localDocumentId>::MaterializeExecutor(
-    MaterializeExecutor<localDocumentId>::Fetcher& /*fetcher*/, Infos& infos)
-    : _buffer{infos.query().resourceMonitor()},
-      _trx{infos.query().newTrxContext()},
-      _readCtx{infos} {
-  if constexpr (localDocumentId) {
-    _collection = infos.collection()->getCollection()->getPhysical();
+std::tuple<ExecutorState, MaterializeStats, AqlCall>
+MaterializeRocksDBExecutor::produceRows(AqlItemBlockInputRange& inputRange,
+                                        OutputAqlItemRow& output) {
+  MaterializeStats stats;
+
+  AqlCall upstreamCall{};
+  upstreamCall.fullCount = output.getClientCall().fullCount;
+
+  auto docRegId = _infos.inputNonMaterializedDocRegId();
+  auto docOutReg = _infos.outputMaterializedDocumentRegId();
+  while (inputRange.hasDataRow() && !output.isFull()) {
+    bool written = false;
+    auto const [state, input] =
+        inputRange.nextDataRow(AqlItemBlockInputRange::HasDataRow{});
+
+    LocalDocumentId id{input.getValue(docRegId).slice().getUInt()};
+    written =
+        _collection
+            ->lookup(
+                &_trx, id,
+                {[&, &input = input](LocalDocumentId /*id*/, VPackSlice doc) {
+                   TRI_ASSERT(input.isInitialized());
+                   output.moveValueInto(docOutReg, input, doc);
+                   return true;
+                 },
+                 [&, &input = input](LocalDocumentId /*id*/,
+                                     std::unique_ptr<std::string>& doc) {
+                   TRI_ASSERT(input.isInitialized());
+                   output.moveValueInto(docOutReg, input, &doc);
+                   return true;
+                 }},
+                {})
+            .ok();
+
+    TRI_ASSERT(written);
```

I think this can be false for rocksdb sometime 
@jsteemann What do you think?

---

If it fails, then something is seriously wrong. If it's a general IO read error, then the query should be aborted, and not silently drop that row from the result set.

---

Sounds good to me.